teps/0082-workspace-hinting.md

```diff
@@ -0,0 +1,303 @@
+performance requirements.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+
+If it's helpful to include workflow diagrams or any other related images,
+add them under "/teps/images/". It's upto the TEP author to choose the name
+of the file, but general guidance is to include at least TEP number in the
+file name, for example, "/teps/images/NNNN-workflow.jpg".
+-->
+
+## Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+
+## Design Evaluation
+<!--
+How does this proposal affect the reusability, simplicity, flexibility 
+and conformance of Tekton, as described in [design principles](https://github.com/tektoncd/community/blob/master/design-principles.md)
+-->
+
+## Drawbacks
+
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
+
+At this stage in the proposal we're just capturing some options to consider. As we move to implementable we'll settle this design on one of them and flesh it out more fully.
+
+### Embed Default Workspace Bindings in Tasks/Pipelines
+
+cf. [Pipelines#4083](https://github.com/tektoncd/pipeline/issues/4083#issuecomment-884874548)
+
+Allow Task and Pipeline authors to explicitly declare some or all of a default Workspace Binding which PipelineRuns and TaskRuns can use or override:
+
+```yaml
+kind: Task
+spec:
+  workspaces:
+  - name: docker-json
+    mountPath: /wherever/docker/json/goes
+    default:
+      secret:
+        secretName: my-docker-json
+```
+
+A TaskRun referencing this Task could either provide a `docker-json` Workspace or omit it. If omitted the Task's default would be used.
+
+A Pipeline could take a similar approach with a `volumeClaimTemplate`:
+
+```yaml
+kind: Pipeline
+spec:
+  workspaces:
+  - name: shared-data
+    default:
+      volumeClaimTemplate:
+        spec:
+          accessModes:
+            - ReadWriteOnce
+          resources:
+            requests:
+              storage: 256M
+```
+
+#### Pros:
+- Very explicit about the Task's expectations for the content provided.
+- Pipelines can override a Tasks' expected types - so, for example, a Task might expect a ConfigMap but a Pipeline might override with a PV instead.
```

Comment:
maybe another pro: adding defaults wouldn't preclude us from adding more explicit hinting later - i.e. there are benefits to providing this functionality in general (e.g. a pipeline that just _works_ out of the box without having to deal with workspaces), AND if this isn't enough we could still add explicit hints (and/or even enforcement) later without needing to undo any of this?

Reply:
I'll add this, thanks.